README
MooX-Role-Modulino

This Role was origianlly written to help standardize command line scripts
at my place of work at the time and to improve testability.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc MooX::Role::Modulino

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        https://rt.cpan.org/NoAuth/Bugs.html?Dist=MooX-Role-Modulino

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/MooX-Role-Modulino

    CPAN Ratings
        https://cpanratings.perl.org/d/MooX-Role-Modulino

    Search CPAN
        https://metacpan.org/release/MooX-Role-Modulino


LICENSE AND COPYRIGHT

Copyright (C) 2019 Jim Bacon

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for MooX-Role-CliOptions

0.04.1_004 2019-09-23 22:35:00
        Added non-modulino style script to examples to verify it works
        from package 'main'. Tweaked POD. Added POD spell-check.

0.04.1_003 2019-09-23 11:11:00
        Set plan to 'skip_all' if it appears that the spawned child process
        fails to run the example script properly. This appears to happen on
        several test platforms even though required modules are specified in
        Makefile.PL.

0.04.1_002 2019-09-22 23:45:00
        Added TEST_REQUIRES to Makefile.PL after see test failures indicating
        that Moo was not available.

0.04.1_001 2019-09-22 23:40:00
        Re-instated tests for moodulino.pl back after adding additional tests
        to help diagnose why they fail on some platforms. Do NOT merge this
        branch to master until that is solved.

0.04    2019-09-21 21:53:00
        Removed tests for moodulino.pl since they do not test the Role
        itself and system() behaves erratically on some systems.

0.03.1_001 2019-09-21 20:13:00
        Added Test::MockModule to build requirements. Investigating test
        failures with system() on some platforms.

0.03    2019-09-21 16:53:00
        Corrected version number in POD.

0.02    2019-09-21 16:44:00
        Removed dependency on '//=' operator.

0.01    2019-09-21 16:00:00
        First version, released on an unsuspecting world.



Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker 6.64;

WriteMakefile(
    NAME               => 'MooX::Role::CliOptions',
    AUTHOR             => q{Jim Bacon <boftx@cpan.org>},
    VERSION_FROM       => 'lib/MooX/Role/CliOptions.pm',
    ABSTRACT_FROM      => 'lib/MooX/Role/CliOptions.pm',
    LICENSE            => 'artistic_2',
    PL_FILES           => {},
    MIN_PERL_VERSION   => '5.006',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '6.64',
    },
    TEST_REQUIRES => {
        'Test::More'              => '0',
        'Test::Exception'         => '0',
        'Test::Deep'              => '0',
        'Test::MockModule'        => '0',
        'Capture::Tiny'           => 0,
        'Moo'                     => 0,
        'MooX::StrictConstructor' => 0,
        'Data::Dumper'            => 0,
    },
    BUILD_REQUIRES => {
        'Test::More'              => '0',
        'Test::Exception'         => '0',
        'Test::Deep'              => '0',
        'Test::MockModule'        => '0',
        'Capture::Tiny'           => 0,
        'Moo'                     => 0,
        'MooX::StrictConstructor' => 0,
        'Data::Dumper'            => 0,
    },
    PREREQ_PM => {
        'Moo'             => 0,
        'Moo::Role'       => 0,
        'Getopt::Long'    => '2.36',
        'Pod::Usage'      => 0,
        'Carp'            => 0,
        'Types::Standard' => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'MooX-Role-CliOptions-*' },
);

__END__


examples/myscript.pl
#!/usr/bin/perl

use strict;
use warnings;

use Data::Dumper;
$Data::Dumper::Indent = 1;

use Moo;
with 'MooX::Role::CliOptions';

use MooX::StrictConstructor;

has custom_opt => ( is => 'ro' );

has internal_attr => (
    is       => 'rw',
    init_arg => undef,
);

my $app = __PACKAGE__->init(
    argv     => \@ARGV,
    add_opts => ['custom_opt=s'],
);
exit $app->run;

# BUILD will be called like normal if present as part of 'init', shown
# here for illustration purposes only.
sub BUILD {
    my $self = shift;

    return;
}

sub run {
    my $self = shift;

    $self->internal_attr('foobar') if !$self->debug;

    print 'custom_opt: ' . $self->custom_opt . "\n" if $self->custom_opt;
    print Dumper($self) if $self->debug;

    return 0;
}

1;
__END__

=pod

=head1 NAME
 
myscript.pl - eample showing how to use MooX::Role::CliOptions
 
=head1 SYNOPSIS
 
myscript.pl [options]
 
 Options:
   --debug    add diagnostic messages and/or disable database writes
   --verbose  extra information in stdout
   --help     brief help message
   --man      full documentation
 
=head1 OPTIONS
 
=over 4
 
=item B<--help>
 
Print a brief help message and exits.
 
=item B<--man>
 
Prints the manual page and exits.
 
=back
 
=head1 DESCRIPTION
 
This script demonstrates how to use C<MooX::Role::CliOptions>. With no
options will print a message and the contents of C<custom_opt>, if any.
 
=head1 AUTHOR

Jim Bacon, C<< <boftx at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Jim Bacon.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut


examples/moodulino.pl
#!/usr/bin/perl

package My::Moodulino;

use strict;
use warnings;

use Data::Dumper;
$Data::Dumper::Indent = 1;

use Moo;
with 'MooX::Role::CliOptions';

use MooX::StrictConstructor;

my $cli = 0;

=pod
has debug => (
    is => 'rw',
    default => 0,
);
=cut

has custom_opt => ( is => 'ro', );

has internal_attr => (
    is => 'rw',

    #init_arg => undef,
);

has cli => (
    is      => 'lazy',
    default => sub { return $cli; },
);

# all attributes and package variable MUST be declared before this!
do {
    print "caller-stack is empty\n";

    # set this so test scripts can see it in the attribute
    $cli = 1;

    my $app = __PACKAGE__->init(
        argv     => \@ARGV,
        add_opts => ['custom_opt=s'],
    );

    exit $app->run;
} unless caller();

# executable code like this should not be used in actual scripts, this
# is only here to provide data for the tests.
print "command line flag not set\n" if !$cli;
print "exit was not called\n";

# BUILD will be called like normal if present as part of 'init', shown
# here for illustration purposes only.
sub BUILD {
    my $self = shift;

    print "cli: $cli\n" if $self->verbose;
}

sub run {
    my $self = shift;

    print "running from command line\n" if $self->cli;
    print 'custom_opt: ' . $self->custom_opt . "\n" if $self->custom_opt;

    print Dumper($self) if $self->debug;

    return 0;
}

1;
__END__

=pod

=head1 NAME
 
moodulino - eample showing how to use MooX::Role::CliOptions
 
=head1 SYNOPSIS
 
moodulino [options]
 
 Options:
   --debug    add diagnostic messages and/or disable database writes
   --verbose  extra information in stdout
   --help     brief help message
   --man      full documentation
 
=head1 OPTIONS
 
=over 4
 
=item B<--help>
 
Print a brief help message and exits.
 
=item B<--man>
 
Prints the manual page and exits.
 
=back
 
=head1 DESCRIPTION
 
This script demonstrates how to use C<MooX::Role::CliOptions>. With no
options will print a message and the contents of C<custom_opt>, if any.
 
=head1 AUTHOR

Jim Bacon, C<< <boftx at cpan.org> >>

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Jim Bacon.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut


lib/MooX/Role/CliOptions.pm
package MooX::Role::CliOptions;

use 5.006;
use strict;
use warnings;

use Carp qw( croak );

use Getopt::Long v2.36 qw(GetOptionsFromArray);
use Pod::Usage;

# this could have been replaced with hand-rolled "isa" clauses, but if
# one is using Moo then the odds are good that one already has this
use Types::Standard qw( ArrayRef Bool );

use Moo::Role;

our $VERSION = '0.04.1_004';

my @options = ( 'help', 'man' );
do {
    has debug => (
        is      => 'ro',
        isa     => Bool,
        default => 1,
    );

    has verbose => (
        is      => 'ro',
        isa     => Bool,
        default => 1,
    );

    push( @options, qw( debug! verbose! ) );
} unless $ENV{MRC_NO_STDOPTS};

has argv => (
    is  => 'ro',
    isa => ArrayRef,
);

sub init {
    my $class = shift;
    my %args  = @_;

    my $argv = delete( $args{argv} ) or croak q{'argv' argument is required};
    croak q{'argv' must be an array reference'}
      unless ref($argv) && ref($argv) eq 'ARRAY';

    my $add_opts = delete( $args{add_opts} ) || [];
    croak q{'add_opts' must be an array reference'}
      unless ref($add_opts) && ref($add_opts) eq 'ARRAY';

    croak q{unknown argument supplied for 'init'} if keys(%args);

    push( @options, $_ ) for ( @{$add_opts} );

    my %values;
    GetOptionsFromArray( $argv, \%values, @options ) or _pod2usage(2);
    _pod2usage(1) if $values{help};
    _pod2usage( -exitstatus => 0, -verbose => 2 ) if $values{man};

    $values{argv} = $argv;

    do {
        # have (no)debug imply verbose if that is not not specificed.
        # This is a # personal preference based on my experience of how
        # these options are most commonly used.
        $values{verbose} ||= $values{debug}
          if defined( $values{debug} ) && !defined( $values{verbose} );
    } unless $ENV{MRC_NO_STDOPTS};

    my $app = eval { $class->new(%values); };
    do {
        print $@ . "\n";
        _pod2usage(2);
    } if $@;

    return $app;
}

# this is needed so test scripts can intercept the call to pod2usage and
# prevent the exit from happening
sub _pod2usage {
    pod2usage(@_);
}

1;    # End of MooX::Role::CliOptions
__END__

=pod

=head1 NAME

MooX::Role::CliOptions - Wrapper to simplify using Moo with Getopt::Long

=head1 VERSION

Version 0.04.1_004

=head1 SYNOPSIS

This is a minimal script that composes C<MooX::Role::CliOptions>.
  
    #!/usr/bin/perl
    package My::Moodulino;
  
    use Moo;
    with 'MooX::Role::CliOptions';
  
    # initialize script attributes/variables
    has option1 => (
        is      => 'ro',
        isa     => sub {
            die 'Illegal value for option1'
              unless $_[0] =~ /^foo|bar$/;
        },
        default => '',
    );
  
    ...
  
    # this makes the script a modulino
    do {
        my $app = __PACKAGE__->init(
            argv     => \@ARGV,
            add_opts => [ 'option1=s' ],
        );
        exit $app->run;
    } unless caller();
  
    sub run {
        my $self = shift;
  
        ...
  
        return 0;
    }
  
    1; # must be present to satisfy "require" when testing
    __END__
  
=head1 DESCRIPTION

This role was written to help standardize command line script behavior and,
if written as a modulino, greatly improve testability. It should be noted
that all example code snippets below are based on the modulino style but
the normal, non-modulino style is also supported. See the example and test
scripts in the distribution for further info if needed.

The default object created when composing this Role has the following
structure (assuming a package name of C<My::Moodulino> as above and no
additional attributes are defined in the composing class.)
  
 $VAR1 = bless( {
  'debug' => 1,
  'verbose' => 1,
  'argv' => []
 }, 'My::Moodulino' );
  
=head1 EXPORTS

None. (Not applicable in a Role.)

=head1 ATTRIBUTES

Three read-only attributes are provided for the composing package, two of
which are exposed as command line arguments and the third being created from
the final C<@ARGV> values after processing by C<Getopt::Long>.

NOTE: Both C<debug> and C<verbose> can be eliminated from the Role by setting
the environment variable C<MRC_NO_STDOPTS> to a "true" value B<before>
composing C<MooX::Role::CliOptions> into your script. This allows you to
completely remove them or redefine them to suit your needs as desired. For
example, you might want C<--debug> to be "off" by default. If you decide to
redefine them you must supply a suitable attribute in your script. This does
NOT affect the C<--help> or C<--man> command line options or the C<argv>
attribute.
  
The following example demonstrates this:
  
 #!perl!
  
 # redefine --debug to be off by default and --verbose to indicate a level
 # between 0 and 5
 BEGIN {
    $ENV{MRC_NO_STDOPTS} = 1;
 }
  
 use Moo;
 with 'MooX::Role::CliOptions';
  
 # not required but strongly recommended by ths author
 use MooX::StrictConstructor;

 has debug => (
    is => 'ro',
    default => 0,
 );
  
 has verbose => (
    is => 'ro',
    isa => sub {
        die 'illegal value for --verbose'
          unless $_[0] >= 0 && $_[0] <= 5;
    },
    default => 0,
 );
  
 do {
    my $app = __PACKAGE__->init(
        argv => \@ARGV,
        add_opts => [
            'debug',
            'verbose=i',
        ],
    );
  
    exit $app->run;
 } unless caller();
  
 ...
  
=head2 debug (Boolean read-only)

Exposed as the negatable C<--debug> command line option.

Default: Boolean TRUE (1). Is turned off with C<--nodebug> on the command
line. (Paranoia is your friend.)

Commonly used to enable diagnostic reports and/or disable potentially
dangerous operations such as database modifications.

Note: Implies the setting for C<verbose> if C<--verbose> or C<--noverbose> is
not explicitly set on the command line.

=head2 verbose (Boolean read-only)

Exposed as the negatable C<--verbose> command line option.

Typically used to add extra information to the output. Often used in
conjunction with C<--debug>.

Default: Will be the same as C<debug> if not explicitly set with either
C<--verbose> or C<--noverbose>. This behavior was chosen since that is
the most common usage pattern in the author's experience.

The most likely usage patterns would be
  
 # in a crontab where no verbose output would be desired (verbose will
 # default to OFF.)
 /my_script --nodebug
  
 # manually run from the cli where normal operation is needed and verbose
 # output is desired
 /my_script --nodebug --verbose

=head2 argv (read-only)

Returns an arrayref to the contents of C<@ARGV> after processing by
C<Getopt::Long>. This is syntactic sugar for using C<@ARGV> directly.

Default: An empty arrayref if no command line arguments other than options
recognized by C<Getopt::Long> are supplied.

=head1 OTHER COMMAND LINE OPTIONS

The following command line options are also accepted, but are not
associated with an attribute since both will cause an immediate exit via
C<Pod::Usage> after displaying the appropriate message.

=head2 --help

Will use C<pod2usage> to display the C<SYNOPSIS> or C<USAGE> POD section if
available.

=head2 --man

Will use C<pod2usage> to display the full POD if available.

=head1 METHODS

=head2 init

This is the workhorse that integrates C<Getopt::Long> with the call to the
composing package's C<new> constructor. Its return value is the resulting
object of the composing package type.

=head3 Parameters

=over 4

=item argv (required)

Typically passed in as follows:
  
 my $app = __PACKAGE__->init( argv => \@ARGV );
  
This will be passed to C<Getopt::Long> for processing. Any remaining elements
will be left in C<@ARGV>. They can also be accessed via C<$app-E<gt>argv>.

=item add_opts (optional)

You can add your own command line options by using this. Simply place the
C<Getopt::Long> specification for any additional options that you want in an
array ref as shown below and be sure to declare an attribute to hold the
option data as processed by C<Getopt::Long>. (See the example scripts to make
things clear.)
  
 # in your script
 has custom1 => (
    is => 'ro',
    isa => sub { ... },
 );
  
 do {
    my $app = __PACKAGE__->init(
        argv => \@ARGV,
        add_opts => [
            'custom1=s',
        ],
    );
    exit ($app->run || 0);
 } unless caller();
  
The above snippet would tell C<Getopt::Long> to accept a command line option
named C<custom1> that must be a string and place it in the C<custom1>
attribute that was declared. You may use any kind of C<isa> or C<coerce>
test that you deem are needed as well as a default.

=back

=head1 SEE ALSO

=over 4

=item C<examples/moodulino.pl> for a functional modulino script.

=item C<examples/myscript.pl> for a functional non-modulino script.

=back

=head1 AUTHOR

Jim Bacon, C<< <boftx at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-moox-role-clioptions at rt.cpan.org>, or through
the web interface at L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooX-Role-CliOptions>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooX::Role::CliOptions


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=MooX-Role-CliOptions>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooX-Role-CliOptions>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/MooX-Role-CliOptions>

=item * Search CPAN

L<https://metacpan.org/release/MooX-Role-CliOptions>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Jim Bacon.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut


lib/MooX/Role/Modulino.pm
package MooX::Role::Modulino;

use 5.006;
use strict;
use warnings;

use Carp qw( croak );

use Getopt::Long v2.36 qw(GetOptionsFromArray);
use Pod::Usage;

use Types::Standard qw( ArrayRef Bool );

use Moo::Role;

our $VERSION = '0.01';

my @options = ( 'help', 'man' );
do {
    has debug => (
        is      => 'ro',
        isa     => Bool,
        default => 1,
    );

    has verbose => (
        is      => 'ro',
        isa     => Bool,
        default => 0,
    );

    push( @options, qw( debug! verbose! ) );
} unless $ENV{MRM_NO_STDOPTS};

has argv => (
    is  => 'ro',
    isa => ArrayRef,
);

sub init {
    my $class = shift;
    my %args  = @_;

    my $argv = delete( $args{argv} ) or croak q{'argv' argument is required};
    croak q{'argv' must be an array reference'}
      unless ref($argv) && ref($argv) eq 'ARRAY';

    my $add_opts = delete( $args{add_opts} ) || [];
    croak q{'add_opts' must be an array reference'}
      unless ref($add_opts) && ref($add_opts) eq 'ARRAY';

    croak q{unknown argument supplied for 'init'} if keys(%args);

    push( @options, $_ ) for ( @{$add_opts} );

    my %values;
    GetOptionsFromArray( $argv, \%values, @options ) or _pod2usage(2);
    _pod2usage(1) if $values{help};
    _pod2usage( -exitstatus => 0, -verbose => 2 ) if $values{man};

    $values{argv} = $argv;

    my $app = eval { $class->new(%values); };
    do {
        print $@ . "\n";
        _pod2usage(2);
    } if $@;

    return $app;
}

# this is needed so test scripts can intercept the call to pod2usage and
# prevent the exit from happening
sub _pod2usage {
    pod2usage(@_);
}

1;    # End of MooX::Role::Modulino
__END__

=pod

=head1 NAME

MooX::Role::Modulino - Wrapper to simplify using Moo with Getopt::Long

=head1 VERSION

Version 0.01

=head1 SYNOPSIS

This is a minimal script that composes C<MooX::Role::Modulino>.
  
    #!perl!
    package My::Modulino;
  
    use Moo;
    with 'MooX::Role::Modulino';
  
    # initialize script attributes/variables
    has option1 => (
        is => 'ro',
        isa => sub {
            die 'Illegal value for option1'
              unless $_[0] =~ /^foo|bar$/;
        },
    );
  
    ...
  
    # this is what makes the script a modulino
    do {
        my $app = __PACKAGE__->init(
            argv     => \@ARGV,
            add_opts => [ 'option1=s' ],
        );
        exit $app->run;
    } unless caller();
  
    sub run {
        my $self = shift;
  
        ...
  
        return 0;
    }
  
    1; # must be present to satisfy "require" when testing
    __END__
  
=head1 DESCRIPTION

This role was written to help standardize command line script behavior and
greatly improve their testability.

=head1 EXPORTS

None.

=head1 ATTRIBUTES

Three attributes are provided for the composing package, two of which are
accepted as command line arguments and the third being created from the final
C<@ARGV> values after processing by C<Getopt::Long>.

NOTE: Both C<--debug> and C<--verbose> can be disabled by setting the
environment variable C<MRM_NO_STDOPTS> to a "true" value B<before> composing
C<MooX::Role::Modulino> into you script. This allows you to completely remove
them or redefine them to suit your needs if so desired. I.e. you might want
C<--debug> to be "off" by default. If you decide to redefine them you must
supply a suitable attribute in your script. This does NOT affect C<--help>
or C<--man>.
  
 #!perl!
  
 # redefine --debug to be off by default and --verbose to indicate a level
 # between 0 and 5
 BEGIN {
    $ENV{MRM_NO_STDOPTS} = 1;
 }
  
 has debug => (
    is => 'ro',
    default => 0,
 );
  
 has verbose => (
    is => 'ro',
    isa => sub {
        die 'illegal value for --verbose'
          unless $_[0] >= 0 && $_[0] <= 5;
    },
    default => 0,
 );
  
 use Moo;
 with 'MooX::Role::Modulino';
  
 do {
    my $app = __PACKAGE__->init(
        argv => \@ARGV,
        add_opts => [
            'debug',
            'verbose=i',
        ],
    );
  
    exit $app->run;
 } unless caller();
  
 ...
  
=head2 debug (Boolean read-only)

Commonly used to either enable extra reports and/or disable potentially
dangerous operations such as database modifications.

Default: ON. Is turned off with C<--nodebug> on the command line.

=head2 verbose (Boolean read-only)

Tyically used to add extra information to the output. Often used in
conjunction with CL--debug>.

Default: OFF. Is turned on with C<--verbose> on the command line.

=head2 argv (read-only)

Returns an arrayref to the contents of C<@ARGV> after processing by
C<Getopt::Long>.

=head1 OTHER COMMAND LINE ARGUMENTS

=head2 --help

Will use C<pod2usage> to display the C<SYNOPSIS> or C<USAGE> POD section if
available.

=head2 --man

Will use C<pod2usage> to display the full POD if available.

=head1 METHODS

=head2 init

This is the workhorse that performs necessary actions and makes the actual
call to the composing package's C<new> constructor. Its return value is
the resulting object of the composing package type.

=head3 Parameters

=over 4

=item argv (required)

Typically passed in as follows:
  
 my $app = __PACKAGE__->init( argv => \@ARGV );
  
This will be passed to C<Getopt::Long> for processing. Any remaining elements
will be left in C<@ARGV>. They can also be accessed via C<$app-E<gt>argv>.

=item add_opts (optional)

You can add your own commandline options by using this. Simply place the
specifications for any additional options that you want in an array ref as
shown below and be sure to declare an attribute to hold the option data
as processed by C<Getopt::Long>. (See the example script to make things clear.)
  
 # in your script
 has custom1 => (
 );
  
 do {
    my $app = __PACKAGE__->init(
        argv => \@ARGV,
        add_opts => [
            'custom1=s',
        ],
    );
    exit ($app->run || 0);
 } unless caller();
  
The above snippet would tell C<Getopt::Long> to accept a command line option
named C<custom1> that must be a string and place it in the C<custom1>
attribute that was declared. You may use any kind of C<isa> or C<coerce>
test that you deem are needed as well as a default.

=back

=head1 SEE ALSO

A fully functional script, C<examples/moodulino.pl>.

=head1 AUTHOR

Jim Bacon, C<< <boftx at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-moox-role-modulino at rt.cpan.org>, or through
the web interface at L<https://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooX-Role-Modulino>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooX::Role::Modulino


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<https://rt.cpan.org/NoAuth/Bugs.html?Dist=MooX-Role-Modulino>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooX-Role-Modulino>

=item * CPAN Ratings

L<https://cpanratings.perl.org/d/MooX-Role-Modulino>

=item * Search CPAN

L<https://metacpan.org/release/MooX-Role-Modulino>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2019 Jim Bacon.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut
